### PR TITLE
[AIRFLOW-2016] Add support for Dataproc Workflow Templates

### PR DESCRIPTION
**Pull request by DanSedov**
Make sure you have checked _all_ steps below.

### JIRA
- [X] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-2016

### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:
Add new operators to support instantiation of Google Cloud Dataproc Workflow Templates.

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
2 new tests in tests/contrib/hooks/test_gcp_dataproc_hook.py

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

- [X] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: tests/contrib/operators/test_dataproc_operator.py
Comment (mbrukman) [inline tests/contrib/operators/test_dataproc_operator.py]: Drop space after `'done'` for consistency.

File: tests/contrib/operators/test_dataproc_operator.py
Comment (mbrukman) [inline tests/contrib/operators/test_dataproc_operator.py]: Add spaces around `=` for consistency.

File: tests/contrib/operators/test_dataproc_operator.py
Comment (mbrukman) [inline tests/contrib/operators/test_dataproc_operator.py]: Add spaces around `=` for consistency.

File: airflow/contrib/operators/dataproc_operator.py
Comment (mbrukman) [inline airflow/contrib/operators/dataproc_operator.py]: Python style is to prefer single-quotes over double-quotes, and you use single-quotes elsewhere for dictionaries.

File: airflow/contrib/operators/dataproc_operator.py
Comment (mbrukman) [inline airflow/contrib/operators/dataproc_operator.py]: Single quotes?

File: airflow/contrib/operators/dataproc_operator.py
Comment (mbrukman) [inline airflow/contrib/operators/dataproc_operator.py]: Single quotes?

**Review by mbrukman — COMMENTED** [review]
Just a few minor stylistic comments; feel free to blame @evilsoapbox for sharing a link to this PR on LinkedIn *and* Twitter. :-)

**Comment by Fokko** [comment]
Looks good. Thanks @mbrukman for the feedback on the style. Also the Flake8 test is failing.

I have one question, what are the changes to the docstring? This probably give conflicts with other merge requests.

**Comment by criccomini** [comment]
@fenglu-g FYI

**Comment by DanSedov** [comment]
@mbrukman, @Fokko Done

File: airflow/contrib/hooks/gcp_dataproc_hook.py
Comment (mbrukman) [inline airflow/contrib/hooks/gcp_dataproc_hook.py]: This should be on the previous line, according to https://www.python.org/dev/peps/pep-0257/#one-line-docstrings — this is a one-line docstring. Similarly elsewhere for single-line docstrings.

File: tests/contrib/operators/test_dataproc_operator.py
Comment (mbrukman) [inline tests/contrib/operators/test_dataproc_operator.py]: This is clear from the name of the class itself, no? You can drop this comment.

**Review by mbrukman — COMMENTED** [review]
Note that you're contributing from your own fork's `master` branch; in general, it's recommended to use a feature branch rather than `master`, which should always stay clean and match the upstream `master`. It also makes it easy to squash and rebase on upstream, etc.

Also, after this PR is merged, you will likely have to delete your fork and re-create it, because your `master` will no longer match the upstream `master` and may or may not be able to sync, whereas with a feature branch, you would just delete it after it is merged, and create a new feature branch from `master`.

Note that if you'd like, you can update this PR to use a different branch in your fork:
https://help.github.com/articles/changing-the-base-branch-of-a-pull-request/

Or you can keep going as-is for this PR, and use a new feature branch for each future PR.

Just a thought.

**Comment by DanSedov** [comment]
@mbrukman Done, also fixed flake8 test

This is my first real experience working with github, I would like to proceed using master for this PR and will do as you suggest for any followups.

**Comment by mbrukman** [comment]
Thanks, @DanSedov! Stylistically seems fine to me now (let's see what Travis CI says).

You'll need a code owner to approve and merge this PR.

File: airflow/contrib/hooks/gcp_dataproc_hook.py
Comment (fenglu-db) [inline airflow/contrib/hooks/gcp_dataproc_hook.py]: Is it possible to revise add a version param to DataprocHook instead of a new hook?

**Comment by DanSedov** [comment]
@fenglu-g Done

File: airflow/contrib/operators/dataproc_operator.py
Comment (fenglu-db) [inline airflow/contrib/operators/dataproc_operator.py]: Do you want to add any template_fields into your operator? maybe template_id, region?

File: airflow/contrib/operators/dataproc_operator.py
Comment (fenglu-db) [inline airflow/contrib/operators/dataproc_operator.py]: There's quite substantial code duplication between templateinstantiate and templateinstantiateinline operators, do you want to refactor and clean up a bit? I'll leave this to you.

File: airflow/contrib/operators/dataproc_operator.py
Comment (DanSedov) [reply airflow/contrib/operators/dataproc_operator.py]: Added template_id

File: airflow/contrib/operators/dataproc_operator.py
Comment (DanSedov) [reply airflow/contrib/operators/dataproc_operator.py]: Could you suggest how this should be merged?

**Comment by DanSedov** [comment]
@fenglu-g Done, added DataprocWorkflowTemplateBaseOperator

**Review by fenglu-db — APPROVED** [review]
LGTM @criccomini please help to merge this PR, thanks.

**Comment by criccomini** [comment]
@fenglu-g Thanks for the review!

LGTM

**Comment by criccomini** [comment]
Merged.

**Pull request closed without merging**
